velox/docs/develop/simd.rst

```diff
@@ -0,0 +1,179 @@
+===================
+SIMD Usage in Velox
+===================
+
+SIMD uses special registers in CPU to operate on multiple primitive data at same
+time.  In some basic cases compiler is able to translate a tight loop into SIMD
+instructions for us, but in some cases we need to call the SIMD intrinsics
+explicitly.
+
+There are several places in Velox where we use SIMD explicitly to get better
+performance.  We also use ``xsimd`` to build zero-cost abstraction over the
+intrinsics, to address portability issue.
+
+Architectures
+-------------
+
+In Velox we support 2 families of CPU architectures regarding SIMD: X86 and ARM.
+In X86 there are 3 generations of SIMD technologies: SSE, AVX, and AVX512.  For
+ARM there are NEON and SVE.  Each architecture has its own size of registers, we
+summarize the details below:
+
+============ ==================== ==================
+Architecture Register Size (bits) Supported in Velox
+============ ==================== ==================
+SSE          128                  Yes
+AVX          256                  Yes
+AVX512       512                  No
+NEON         128                  Yes
+SVE          128 - 2048           No
+============ ==================== ==================
+
+xsimd Basics
+------------
+
+The data structure in ``xsimd`` to represent a SIMD register is ``batch<T, A>``.
+``T`` stands for the element data type and ``A`` stands for the architecture.
+This type has only 1 field called ``data``, which is the underlying SIMD
+register (e.g. for AVX it can be ``__m256``, ``__m256d``, or ``__m256i``).  This
+ensures the data structure can be optimized directly as the register without any
+overhead during runtime.
+
+When you compare 2 SIMD vectors (e.g. ``x == y``), there are 2 kinds of result
+type, depending on the architecture.  For AVX512, the comparison result is kept
+as a bit mask (1 bit per element, up to 64 bits) in a normal integer.  For all
+other architectures, the result is kept in another SIMD register with the same
+number of lanes as the operands, and each element in the result is set to
+either -1 (``true``) or 0 (``false``).  In ``xsimd`` these 2 types are unified
+to one type: ``batch_bool<T, A>``.  ``T`` is the element type of comparison
+operands, and ``A`` is the architecture.
+
+``xsimd`` provides some functions and opeartors to abstract the intrinsics on
+different architectures, including basic arithmetics, comparisons, bitwise
+opeartions, mathematical functions, loading or storing from memory.
+
+SIMD Utilities
+--------------
+
+There are some intrinsics that is not yet abstracted by ``xsimd``.  We add the
+common ones in ``common/base/SimdUtil.h``.
```

Comment:
Do you plan to submit changes upstream?

Reply:
Not for now.  We are using bit mask heavily in these extensions, and in some cases is not zero cost (optimal in our cases though), not sure if that aligns with the philosophy of the library.